Server/src/RedisManager.cpp

```diff
@@ -0,0 +1,236 @@
+ *
+ * \param [in] logPtr     logger pointer
+ ***********************************************************************/
+void RedisManager::Setup(Logger *logPtr, BMPListener::ClientInfo *client) {
+    logger = logPtr;
+    client_ = client;
+}
+
+
+
+/**
+ * Get Key separator for deletion
+ *
+ * \param [in] N/A
+ */
+std::string RedisManager::GetKeySeparator() {
+    return separator_;
+}
+
+
+/**
+ * WriteBMPTable
+ *
+ * \param [in] table            Reference to table name
+ * \param [in] key              Reference to various keys list
+ * \param [in] fieldValues      Reference to field-value pairs
+ */
+bool RedisManager::WriteBMPTable(const std::string& table, const std::vector<std::string>& keys, const std::vector<swss::FieldValueTuple> fieldValues) {
+
+    if (enabledTables_.find(table) == enabledTables_.end()) {
+        LOG_INFO("RedisManager %s is disabled", table.c_str());
+        return false;
+    }
+
+    swss::Table stateBMPTable(&stateDb_, table);
+    std::string fullKey;
+    for (const auto& key : keys) {
+        fullKey += key;
+        fullKey += separator_;
+    }
+    fullKey.erase(fullKey.size() - 1);
+
+    LOG_INFO("RedisManager WriteBMPTable key = %s", fullKey.c_str());
+
+    stateBMPTable.set(fullKey, fieldValues);
+    return true;
+}
+
+
+/**
+ * RemoveBMPTable
+ *
+ * \param [in] keys             Reference to various keys
+ */
+bool RedisManager::RemoveBMPTable(const std::vector<std::string>& keys) {
+
+    stateDb_.del(keys);
+    return true;
+}
+
+/**
+ * DisconnectBMP
+ *
+ * \param [in] N/A
+ */
+void RedisManager::DisconnectBMP() {
+    LOG_INFO("RedisManager DisconnectBMP");
+    close(client_->c_sock);
+    client_->c_sock = 0;
+}
+
+/**
+ * ExitRedisManager
+ *
+ * \param [in] N/A
+ */
+void RedisManager::ExitRedisManager() {
+    exit_ = true;
+    for (auto& threadPtr : threadList_) {
+        threadPtr->join();
+    }
+}
+
+/**
+ * ReadBMPTable, there will be dedicated thread be launched inside and monitor corresponding redis table.
+ *
+ * \param [in] tables             table names to be subscribed.
+ */
+void RedisManager::SubscriberWorker(const std::string& table) {
+    try {
+        swss::DBConnector cfgDb("CONFIG_DB", 0, true);
+
+        swss::SubscriberStateTable conf_table(&cfgDb, table);
+        swss::Select s;
+        s.addSelectable(&conf_table);
+
+        while (!exit_) {
+            swss::Selectable *sel;
+            int ret;
+
+            ret = s.select(&sel, BMP_CFG_TABLE_SELECT_TIMEOUT);
+            if (ret == swss::Select::ERROR) {
+                SWSS_LOG_NOTICE("Error: %s!", strerror(errno));
+                continue;
+            }
+            if (ret == swss::Select::TIMEOUT) {
+                continue;
+            }
+
+            swss::KeyOpFieldsValuesTuple kco;
+            conf_table.pop(kco);
+
+            if (std::get<0>(kco) == "SET") {
+                if (std::get<1>(kco) == "true") {
+                    EnableTable(table);
```

Comment:
>EnableTable [](http://example.com/codeflow?start=20&length=11)

`EnableTable(table);` is not enough, you will lose the data before enable that config.

Reply:
changed flow as above comment mentioned.